### PR TITLE
Tutorial introduction to Ignition Physics

### PR DESCRIPTION
**Pull request by anindex**
The initial introduction draft is ready for review. The content is written according to this #103 planning.

### CODE REVIEW

File: tutorials/01_intro.md [hunk 1]
```diff
@@ -2,18 +2,61 @@
...
+## Motivation
+
+Ignition Physics (`ign-physics`) is a library component in Ignition Robotics, a set of libraries designed to rapidly develop robot and simulation applications.
+The main goal of the library is to provide an abstraction layer to various physics engines, which gives end-users the ability to use multiple physics engines with minimal change to their application code.
+Users can request for physics engines that support a set of features and the plugin loading mechanism loads only the engines that implement the requested features.
+Besides, a user-selected physics engine can be integrated into the existing code-base by writing a compatible plugin interface, which enables Ignition Physics extensibility and modularity.  
+
+## Overview
+
+For a big picture of the Ignition Physics operation in Ignition Robotics ecosystem, see the abstract diagram below:
```
Comment (iche033) [inline tutorials/01_intro.md:14]: same as above: `Ignition Robotics` -> `Ignition`

File: tutorials/01_intro.md [hunk 2]
```diff
@@ -2,18 +2,61 @@
...
+
+Ignition Physics uses a plugin architecture where each physics engine is implemented as a plugin that can be loaded at runtime.
+To enable users the ability to choose between physics engines, Ignition Physics introduces four keys elements:
+
+- \ref ignition::physics::Entity "Entity": the base class, which abstracts all proxy objects, which contains minimal descriptive data (e.g. a unique identifier, a reference-counter, etc.) pointed the corresponding implemented objects.
+- \ref ignition::physics::FeaturePolicy "FeaturePolicy": the policy class, which provides metadata to features about their simulation engine specifications.
+- \ref ignition::physics::Feature "Feature": defines the concept used to encode the capabilities of a physics engine.
+- \ref ignition::physics::FeatureList "FeatureList": aggregates a list of features.
+
+Depending on using an external physics engine (e.g. DART, TPE, Bullet, etc.), the internal interface to the physics engine might be different.
```
Comment (iche033) [inline tutorials/01_intro.md:31]: `Depending on which external physics engine is used ...`?

File: tutorials/01_intro.md [hunk 3]
```diff
@@ -2,18 +2,61 @@
...
+- Add features for TPE physics engines (more detail in \ref physicsplugin).
+- Extend contact data with force, normal, and penetration depth.
+- Add Base and EntityManagement to `tpeplugin`
+
+**Ignition Physics 3.x**
+
+### Future roadmap
+
+Ignition Physics evolves closely with the Ignition Robotics ecosystem.
+For a broader overview, please visit [Ignition Robotics's roadmap](https://ignitionrobotics.org/about).
```
Comment (iche033) [inline tutorials/01_intro.md:62]: same as above `Ignition Robotics` -> `Ignition`

File: tutorials/01_intro.md (hunk 2 above)
Comment (anindex) [reply tutorials/01_intro.md:31]: I mean `ign-physics` is an abstraction layer to interface features and functionalities of specific physics engines. Each physics engines have a different API, so interfacing with them is different case by case. Pls correct me if I am wrong :)

File: tutorials/01_intro.md [hunk 4]
```diff
@@ -2,18 +2,61 @@
...
+Ignition Physics (`ign-physics`) is a library component in Ignition, a set of libraries designed to rapidly develop robot and simulation applications.
+The main goal of the library is to provide an abstraction layer to various physics engines, which gives end-users the ability to use multiple physics engines with minimal change to their application code.
+Users can request for physics engines that support a set of features and the plugin loading mechanism loads only the engines that implement the requested features.
+Besides, a user-selected physics engine can be integrated into the existing code-base by writing a compatible plugin interface, which enables Ignition Physics extensibility and modularity.  
+
+## Overview
+
+For a big picture of the Ignition Physics operation in Ignition ecosystem, see the abstract diagram below:
+
+<img src="https://user-images.githubusercontent.com/18066876/94050480-9dc54000-fdd6-11ea-92e7-832dcdf4caba.png"/>
```
Comment (azeey) [inline tutorials/01_intro.md:16]: Thanks for the nice diagram! The part about ign-transport and ign-msgs being used for communication makes it appear that ign-physics communicates, for example, with ign-sensors via ign-transport. However, this is not quite the case. In the current versions of Ingition, only ign-gazebo interfaces with ign-physics via it's C++ API. ign-gazebo received state data from ign-physics and sends them to other libraries either directly or via ign-transport. For the particular libraries in the diagram (ign-sensors and ign-rendering), ign-gazebo sends the information directly via their C++ API.

File: tutorials/01_intro.md [hunk 5]
```diff
@@ -2,18 +2,61 @@
...
+
+In general, `ign-gazebo` is the main simulation library, in which its functionalities are powered by many component libraries.
+For example, its graphical drawing is supported by `ign-rendering` or simulated sensors that are defined and implemented in `ign-sensors`.
+In particular, this library `ign-physics` provides an abstract interface to physics engines, which simulates dynamic transformations and interactions of objects in `ign-gazebo`.
+The communication between these libraries at runtime is provided by `ign-transport` and `ign-msgs`.
+
+Ignition Physics uses a plugin architecture where each physics engine is implemented as a plugin that can be loaded at runtime.
+To enable users the ability to choose between physics engines, Ignition Physics introduces four keys elements:
+
+- \ref ignition::physics::Entity "Entity": the base class, which abstracts all proxy objects, which contains minimal descriptive data (e.g. a unique identifier, a reference-counter, etc.) pointed the corresponding implemented objects.
```
Comment (azeey) [inline tutorials/01_intro.md:26]: Part of this sentence, "pointed the corresponding...", seems like it's missing something. Is "pointed" meant to be "pointing to"?

File: tutorials/01_intro.md [hunk 6]
```diff
@@ -2,18 +2,61 @@
...
-capabilities of a physics engine. Users can request for physics engines
-that support a set of features and the plugin loading mechanism loads only
-the engines that implement the requested features.
-
-The following is a list of currently supported physics engines:
-
-  * [DART](http://dartsim.github.io/)
+## Motivation
+
+Ignition Physics (`ign-physics`) is a library component in Ignition, a set of libraries designed to rapidly develop robot and simulation applications.
```
Comment (azeey) [inline tutorials/01_intro.md:7]: Consider keeping the lines to 80 characters so git can generate better diffs.

File: tutorials/01_intro.md (hunk 2 above)
Comment (scpeters) [reply tutorials/01_intro.md:31]: I think he is just suggesting a change of phrasing; your meaning is correct.

`Depending on using an external physics engine` -> `Depending on which external physics engine is used`

File: tutorials/01_intro.md (hunk 5 above)
Comment (scpeters) [reply tutorials/01_intro.md:26]: "pointed" -> "pointing" sounds right to me

File: tutorials/01_intro.md [hunk 7]
```diff
@@ -2,18 +2,61 @@
...
+In general, `ign-gazebo` is the main simulation library, in which its functionalities are powered by many component libraries.
+For example, its graphical drawing is supported by `ign-rendering` or simulated sensors that are defined and implemented in `ign-sensors`.
+In particular, this library `ign-physics` provides an abstract interface to physics engines, which simulates dynamic transformations and interactions of objects in `ign-gazebo`.
+The communication between these libraries at runtime is provided by `ign-transport` and `ign-msgs`.
+
+Ignition Physics uses a plugin architecture where each physics engine is implemented as a plugin that can be loaded at runtime.
+To enable users the ability to choose between physics engines, Ignition Physics introduces four keys elements:
+
+- \ref ignition::physics::Entity "Entity": the base class, which abstracts all proxy objects, which contains minimal descriptive data (e.g. a unique identifier, a reference-counter, etc.) pointed the corresponding implemented objects.
+- \ref ignition::physics::FeaturePolicy "FeaturePolicy": the policy class, which provides metadata to features about their simulation engine specifications.
```
Comment (scpeters) [inline tutorials/01_intro.md:27]: I would mention the two types of metadata currently supported by the `FeaturePolicy` class:

* Scalar numeric type (64-bit `double` or 32-bit `float`)
* Dimensionality (2D or 3D)

I've seen numerous physics engines that support different scalar types, but I've never seen one that supports both 2D and 3D

File: tutorials/01_intro.md (hunk 4 above)
Comment (anindex) [reply tutorials/01_intro.md:16]: @azeey Thanks for the explanations :) Then I will remove the communication description between `ign-physics` with other libraries.

File: tutorials/01_intro.md (hunk 5 above)
Comment (anindex) [reply tutorials/01_intro.md:26]: ye :) that's a typo

**Review by azeey — APPROVED** [review]

**Review by claireyywang — APPROVED** [review]

**Pull request merged**